### ISSUE TITLE
Transitive TypeScript parser causes peer dependency warning with TypeScript 7

### ISSUE THREAD

**Issue by transitive-bullshit** [shadcn-ui/lint#1]
Installing `@shadcn/lint@0.1.0` in an Oxlint project using TypeScript `7.0.2` succeeds but produces a peer dependency warning.

`@shadcn/lint` depends on `@typescript-eslint/parser`, which resolves to `8.70.0`. Its supported TypeScript range excludes the installed version.

Running `pnpm peers check` reports:

```text
✕ unmet peer typescript
  Installed: 7.0.2
  Wanted:
    ">=4.8.4 <6.1.0":
      @typescript-eslint/parser@8.70.0
      @typescript-eslint/typescript-estree@8.70.0
      @typescript-eslint/tsconfig-utils@8.70.0
      @typescript-eslint/project-service@8.70.0
```

Environment:

- `@shadcn/lint`: 0.1.0
- Oxlint: 1.82.0
- TypeScript: 7.0.2
- pnpm: 12.3.4
- Node.js: 25.9.0

The configuration loads, and all six shadcn rules run and report findings without an observed parser crash. This report concerns the dependency compatibility warning; it does not establish full TypeScript 7 compatibility.

Could you clarify whether TypeScript 7 is supported, and whether the parser dependency can be updated or made optional for Oxlint users?

**Comment by rdjperron** [comment]
I hit the same thing from a different angle and can add what the dependency does to an Oxlint-only pnpm workspace, plus a reproduction of the fix you asked about (making the parser optional).

## What loads the parser

`@typescript-eslint/parser` is only ever loaded lazily, as a fallback, in `packages/lint/src/project/parser.ts`: `createParser()` tries `require("oxc-parser")` first and only calls `loadTypeScript()` (the `require("@typescript-eslint/parser")` at line 64) when that fails. `oxc-parser` is in `optionalDependencies`, which makes sense because it ships platform bindings. So on every platform oxc-parser supports, the TypeScript parser is installed and never required. CONTRIBUTING.md says the same: "With `oxc-parser` installed, this analysis does not load TypeScript under Oxlint."

## What the hard dependency costs

`@typescript-eslint/parser@8.70.0` declares peers `eslint ^8.57.0 || ^9.0.0 || ^10.0.0` and `typescript >=4.8.4 <6.1.0`. Both npm and pnpm auto-install missing non-optional peers, so an Oxlint-only install of `@shadcn/lint@0.1.0` also installs ESLint and a TypeScript, even though neither is used.

Single package, npm 11, `npm install -D @shadcn/lint oxlint tailwindcss`:

```text
node_modules/eslint      10.10.0
node_modules/typescript  6.0.3
```

pnpm 11.22.0 workspace, plugin at the root, Tailwind and `typescript: catalog:` (catalog pins `7.0.2`) only in `apps/web`, `vite-plus@0.3.1` at the root, fresh lockfile:

```yaml
importers:
  .:
    devDependencies:
      '@shadcn/lint':
        version: 0.1.0(eslint@10.10.0)(typescript@6.0.3)
      vite-plus:
        version: 0.3.1(typescript@6.0.3)
  apps/web:
    devDependencies:
      typescript:
        version: 7.0.2
```

`pnpm why eslint` shows the only path is `@shadcn/lint > @typescript-eslint/parser`. The second line is the part that bites: the auto-installed `typescript@6.0.3` becomes the root's TypeScript, so every root package with an optional `typescript` peer (vite-plus's bundled tsdown here) binds to 6.0.3 instead of the workspace's 7.0.2. Declaring `typescript: catalog:` at the root fixes that binding and leaves exactly the warning in this issue:

```text
✕ unmet peer typescript
  Installed: 7.0.2
  Wanted:
    ">=4.8.4 <6.1.0":
      @typescript-eslint/parser@8.70.0
      @typescript-eslint/typescript-estree@8.70.0
      @typescript-eslint/project-service@8.70.0
      @typescript-eslint/tsconfig-utils@8.70.0
```

## A silent failure hiding behind it

Both callers of `parseSource` (`variants.ts:147-151`, `wrappers.ts:201-205`) catch parser errors and return "nothing found". So if neither parser resolves, the rules keep running but lose cross-file analysis with no warning. On 0.1.0 with `oxc-parser` and `@typescript-eslint/parser` both removed from `node_modules`, `<Button className="bg-pink-500">` goes from

```text
"bg-pink-500" is not allowed on <Button>: <Button> owns its color. Use a variant: default, outline, destructive. ...
```

to

```text
"bg-pink-500" is not allowed on <Button>: <Button> owns its color. Add a variant in src/components/ui/button.tsx ...
```

with nothing on stderr. That is the case that a hard dependency currently prevents on platforms without oxc bindings, so making the parser optional should come with a warning for it.

## Proposed change

Move `@typescript-eslint/parser` from `dependencies` to `peerDependencies` with `peerDependenciesMeta: { optional: true }` (keep it in `devDependencies` for the test suite), and have `loadTypeScript()` call `warnOnce` before rethrowing when the module is missing. ESLint users are unaffected: the README's ESLint setup already installs `@typescript-eslint/parser` explicitly, and any project parsing TSX with ESLint has a parser installed that pnpm and npm will resolve as the peer.

I verified this on a build from `main` (53de86f):

- `pnpm lint`, `pnpm typecheck`, `pnpm test`: 52 files, 697 tests pass, 1 skipped.
- The packed tarball in the workspace above: no `eslint`, no `typescript@6.0.3` in the store, `vite-plus` resolves `(typescript@7.0.2)`, `pnpm peers check` no longer lists `@typescript-eslint/parser`.
- Oxlint 1.81.0 with the plugin still reads Button's variants from `components/ui/button.tsx` (through oxc-parser).
- With oxc-parser removed as well, stderr now says:

```text
[@shadcn/lint] No parser for component files: neither oxc-parser nor @typescript-eslint/parser could be loaded, so variants and className forwarding in other files are not seen. Install oxc-parser, or @typescript-eslint/parser.
```

That change is on a branch, with a changeset: https://github.com/rdjperron/shadcn-lint/tree/fix/optional-typescript-parser (diff: https://github.com/shadcn-ui/lint/compare/main...rdjperron:shadcn-lint:fix/optional-typescript-parser). This repository limits pull requests to collaborators, so I cannot open one; happy to if invited, or feel free to pull it in. It does not answer the TypeScript 7 question itself; that stays with typescript-eslint's supported range.

Environment: `@shadcn/lint` 0.1.0, Oxlint 1.81.0, pnpm 11.22.0 and npm 11, Node 26.7.0, TypeScript 7.0.2 (catalog), Tailwind 4.3.3.

---

_This comment was written by Claude Fable 5.1 (Claude Code) on behalf of @rdjperron, who reviewed and verified it before posting._
